src/editorwatchdog.js

```diff
@@ -0,0 +1,323 @@
+	 *
+	 * @method #setCreator
+	 * @param {Function} creator
+	 */
+
+	/**
+	 * Sets the function that is responsible for the editor destruction.
+	 * Overrides the default destruction function, which destroys only the editor instance.
+	 * It expects a function that should return a promise or `undefined`.
+	 *
+	 *		watchdog.setDestructor( editor => {
+	 *			// Do something before the editor is destroyed.
+	 *
+	 *			return editor
+	 *				.destroy()
+	 *				.then( () => {
+	 *					// Do something after the editor is destroyed.
+	 *				} );
+	 *		} );
+	 *
+	 * @method #setDestructor
+	 * @param {Function} destructor
+	 */
+
+	/**
+	 * Restarts the editor instance. This method is called whenever an editor error occurs. It fires the `restart` event and changes
+	 * the state to `initializing`.
+	 *
+	 * @protected
+	 * @fires restart
+	 * @returns {Promise}
+	 */
+	_restart() {
+		return Promise.resolve()
+			.then( () => {
+				this.state = 'initializing';
+
+				return this._destroy();
+			} )
+			.catch( err => {
+				console.error( 'An error happened during the editor destroying.', err );
+			} )
+			.then( () => {
+				if ( typeof this._elementOrData === 'string' ) {
+					return this.create( this._data, this._config, this._config.context );
+				} else {
+					const updatedConfig = Object.assign( {}, this._config, {
+						initialData: this._data
+					} );
+
+					return this.create( this._elementOrData, updatedConfig, updatedConfig.context );
+				}
+			} )
+			.then( () => {
+				this.fire( 'restart' );
+			} );
+	}
+
+	/**
+	 * Creates and keep running the editor instance using the defined creator and destructor.
+	 *
+	 * @param {HTMLElement|String|Object.<String|String>} [elementOrData] Editor's source element or the editor's data.
+	 * @param {module:core/editor/editorconfig~EditorConfig} [config] Editor configuration.
+	 * @param {Object} [context] A context for the editor.
+	 *
+	 * @returns {Promise}
+	 */
+	create( elementOrData = this._elementOrData, config = this._config, context ) {
+		return Promise.resolve()
+			.then( () => {
+				super._startErrorHandling();
+
+				this._elementOrData = elementOrData;
+
+				// Clone configuration because it might be shared within multiple watchdog instances. Otherwise,
+				// when an error occurs in one of these editors, the watchdog will restart all of them.
+				this._config = this._cloneEditorConfiguration( config ) || {};
+
+				this._config.context = context;
+
+				return this._creator( elementOrData, this._config );
+			} )
+			.then( editor => {
+				this._editor = editor;
+
+				this.listenTo( editor.model.document, 'change:data', this._throttledSave );
+
+				this._lastDocumentVersion = editor.model.document.version;
+				this._data = this._getData();
+
+				this.state = 'ready';
+			} );
+	}
+
+	/**
+	 * Destroys the current editor instance by using the destructor passed to the {@link #setDestructor `setDestructor()`} method
```

Comment:
I guess it rather destroys watchdog, so I'd write that it destroys watchdog and the editor instance. I wouldn't write that it uses callback passed to `setDestructor()` or I'd write that it destroys editor instance or fires that callback. I hope you get what I mean.